data-acquisition/data-synthesizer/src/data_synthesizer/data_synthsizer.py

```diff
@@ -0,0 +1,159 @@
+        curr_bg_conf = self.bg_conf[curr_bg_label]
+        curr_bg_file = (
+            self.template_path + "/backgrounds/{bg}.jpg").format(
+                bg=curr_bg_label)
+        return curr_bg_file, curr_bg_conf
+
+    def generate_synthetic_dataset(
+            self, n, categories=["bottles"], rot_pc=0.1, x_offset=1,
+            obj_sizes_allowed=["s", "m", "l"], max_objs_in_pack=3):
+        """
+            The synthesizer takes the following parameters:
+            n: integer specifying the number of images to be generated.
+
+            categories (default: ["bottles"]): An array specifying the 
+                categories from which objects will be selected. Should 
+                match the keys in the foreground config file.
+
+            rot_pc (default=0.1): Proportion of rotated images
+
+            x_offset (default=1 px): Integer value specifying the number of
+                pixels present between two images.
+
+            obj_sizes_allowed (default=[["s", "m", "l"]]): An array specifying
+                the allowed variation in sizes for the objects.
+                (s=small, m=medium and l=large)
+
+            max_objs_in_pack (default=3): Max number of objects in a pack 
+                (appearing consecutively).
+        """
+
+        timestamp = time.strftime("%Y_%m_%d_%H_%M_%S", time.gmtime())
+        dataset_name = "synth_data_{}".format(timestamp)
+        save_path = self.data_path + dataset_name
+        os.mkdir(save_path)
+
+        for i in range(n):
+
+            image_name = "image_{}_{}".format(i, timestamp)
+            image_path = save_path + "/{}".format(image_name)
+            os.mkdir(image_path)
+            os.mkdir(image_path + "/train_image")
+            os.mkdir(image_path + "/train_mask")
+
+            print("Generating image {i} of {n}".format(i=i+1, n=n))
+            bg_file, bg_conf = self._get_bg_and_conf()
+
+            num_shelves = bg_conf["num_shelves"]
+            shelf_region = bg_conf["shelf_region"]
+            x_start, x_end = shelf_region[0][0], shelf_region[1][0]
+            shelf_positions = bg_conf["shelf_y_positions"]
+            shelf_ht = bg_conf["shelf_ht"]
+
+            bg_img = Image.open(bg_file)
+            empty_fr = Image.new('RGBA', bg_img.size, color=(0, 0, 0, 0))
+            empty_alpha_mask = Image.new('L', bg_img.size, color=0)
+            num_obj = 0
+            for shelf in range(num_shelves):
+                paste_pos = (x_start, 0)
+                while paste_pos[0] < x_end:
+
+                    fr_file, fr_conf = self._get_fr_and_conf(categories)
+                    new_size = self._get_new_image_size(
+                        fr_conf, obj_sizes_allowed, shelf_ht)
+                    fr_img = Image.open(fr_file)
+                    fr_img = fr_img.resize(new_size, Image.LANCZOS)
+
+                    num_instances = np.random.choice(max_objs_in_pack, 1)[0]
+
+                    for instance in range(num_instances):
```

Comment:
This a fourth-level nested loop. @pshivraj will agree that this does not sound like a good idea :P Can we please refactor?

Reply:
Yeah, I was apprehensive about this as well. But then resolving this would require design changes which would take more time. So my idea was to give you guys something to start with and then make design changes.

Reply:
Yeah, that makes sense. If this is a blocker for anyone else on the team, please feel free to merge this as is, and we can continue working on this branch and make the required changes.